pkg/storage/unified/sql/backend.go

```diff
@@ -650,15 +650,23 @@ func (b *backend) poll(ctx context.Context, grp string, res string, since int64,
 // in a single roundtrip. This would reduce the latency of the operation, and also increase the
 // throughput of the system. This is a good candidate for a future optimization.
 func resourceVersionAtomicInc(ctx context.Context, x db.ContextExecer, d sqltemplate.Dialect, key *resource.ResourceKey) (newVersion int64, err error) {
-	// TODO: refactor this code to run in a multi-statement transaction in order to minimize the number of round trips.
-	// 1 Lock the row for update
-	rv, err := dbutil.QueryRow(ctx, x, sqlResourceVersionGet, sqlResourceVersionRequest{
+
+	// 1. Increase the resource version to the current epoch with microseconds precision.
+	// This is an update statement that will lock to row and prevent concurrent updates until the transaction is commited.
+	//
+	// Note: there is a small chance that the resource version will be the same as the previous one when
+	// concurrent updates are too close in time. This is due to the fact that the current epoch is unlikely to cause
+	// real issues in practice (given network latency, etc).
+	// In rare occasions (like master clock changes) the resource version might go backwards for a short period of time.
+	// In those scenarios, watchers might not receive the expected events.
```

Comment:
The approach below would work around this too I think, with RVs incrementing until the clock catches up.